### ISSUE TITLE
P1642 Freestanding [utilities], [ranges], and [iterators]

### ISSUE THREAD

**Issue by wg21bot** [cplusplus/papers#429]
[P1642R0](https://wg21.link/p1642r0) Freestanding Library: Easy [utilities] (Ben Craig)

**Comment by wg21bot** [comment]
[P1642R1](https://wg21.link/p1642r1) Freestanding Library: Easy [utilities], [ranges], and [iterators] (Ben Craig)

**Comment by wg21bot** [comment]
[P1642R2](https://wg21.link/p1642r2) Freestanding Library: Easy [utilities], [ranges], and [iterators] (Ben Craig)

**Comment by brycelelbach** [comment]
[Prague 2020-01 LEWGI Minutes](http://wiki.edg.com/bin/view/Wg21prague/P1642)

[P1642R2](https://wg21.link/P1642R2) Freestanding [utilities], [ranges], and [iterators]: Design Review

__Chair:__ Bryce Adelstein Lelbach

__Champion:__ Ben Craig

__Minute Taker:__ Peter Bindels

__Start Review:__ 2020-02-10 11:06

Major design points:
- Splitting `std::swap` overload set: some `std::swap` overloads will be freestanding, some will not.
- Omitting feature test macros.

Should `std::function` be in freestanding?

__End:__ 11:31

__CONSENSUS: LEWGI sends P1642R2 (Freestanding Status Quo) to LEWG.__

**Comment by brycelelbach** [comment]
[P1642R3](https://wg21.link/P1642R0): Freestanding [utilities], [ranges], and [iterators]

[2020-06-09 Library Evolution Telecon Minutes](http://wiki.edg.com/bin/view/Wg21summer2020/P1642#Library-Evolution-2020-06-09)

__Chair:__ Bryce Adelstein Lelbach

__Champion:__ Ben Craig

__Minute Taker:__ Mark Hoemmen

__Start Review:__ 2020-06-09 10:06 Pacific

__POLL: Add `__cpp_lib_freestanding_*` feature test macros (either per header or per feature).__

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
|3|13|5|0|0|

__Attendance:__ 26

__# of Authors:__ 1 

__Author Position:__ WF

That has strong consensus.

__POLL: Add `__cpp_lib_freestanding_*` per-header feature test macros (proposed by P1642R3).__

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
|2|5|11|3|0|

__Attendance:__ 26

__# of Authors:__ 1

__Author Position:__ SF

That does not have consensus.

__POLL: The freestanding versions of <tuple>, <ratio>, and <utility> proposed by P1642R3, excluding feature test macros, are ready to forward to LWG.__

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
|13|7|1|0|0|

__Attendance:__ 26

__# of Authors:__ 1

__Author Position:__ SF

That has strong consensus.

__POLL: The freestanding versions of <ranges>, <memory>, <iterator>, and <functional> proposed by P1642R3, excluding feature test macros, are ready to forward to Library.__

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
|7|12|2|0|0|

__Attendance:__ 26

__# of Authors:__ 1

__Author Position:__ SF

That has strong consensus.

__End:__ 11:40

__SUMMARY:__

We reviewed P1642R3, which proposes adding library features to freestanding that
are relatively straightforward to include. The proposed additions to
freestanding - `<tuple>`, `<utility>` and `<ratio>`, and a subsetted versions of
`<ranges>`, `<memory>`, `<iterator>`, and `<functional>` - were largely
non-controversial and we had strong consensus that these parts were ready to move
forward to Library.

We discussed how feature test macros should work in freestanding at great length.
We agreed that we wanted freestanding-specific feature test macros, although it
is an open question whether said feature test macros should also be defined in
hosted. Additional investigation on freestanding feature test macros is needed.

P1642R3 depends on P1641R3, which adds some changes how the freestanding library
is defined and adds some front matter; we reviewed P1641R3 last week.
We've asked the author to take the freestanding feature test macro content
out of P1641R3 and P1642R3 and put it in a new paper, and then combine P1641R3
and P1642R3 into a single paper, have a wordsmith (Walter Brown) review the
more complicated wording from P1641R3, and then bring the combined paper back
to Library Evolution for a final review before forwarding it to Library.

__CONSENSUS:__
   * Split the feature test macro parts of P1641 and P1642 out into a third
     paper. Explore whether `__cpp_lib_freestanding_*` feature test macros
     should be defined in both hosted and freestanding. Show examples of usage.
   * Combine P1641 and P1642 minus feature test macros into one paper; return to
     Library Evolution with said paper. We will review the non P1642 parts with
     the goal of advancing them to Library.
   * We need to make some policy decisions here at some point, such as how to
     handle partially supported standard library classes in freestanding, how to
     handle partial headers, the rubric for what should be included in
     freestanding. P0829, P1641, and P1642 cover some of this. Consider writing
     a policy paper and proposing a standing document.

**Comment by wg21bot** [comment]
[P1642R4](https://wg21.link/p1642r4) Freestanding Library: Easy [utilities], [ranges], and [iterators] (Ben Craig)

**Comment by brycelelbach** [comment]
[P1642R4](https://wg21.link/P1642R0): Freestanding [utilities], [ranges], and [iterators]

[2020-11-30 Library Evolution Telecon Minutes](http://wiki.edg.com/bin/view/Wg21summer2020/P1642#Library-Evolution-2020-11-30)

__POLL:__ Send P1642R4 (Freestanding [utilities], [ranges], and [iterators]) after rebasing against the current WD and add _Exit to electronic ballot to be forwarded to LWG for C++IS.

|SF|F|N|A|SA|
|-|-|-|-|-|
|5|7 |0|0|0|

__Attendance:__ 19

__Author Position:__ SF

Conclusion: Consensus in favor.

__POLL:__ Which priority bucket should P1642 (Freestanding [utilities], [ranges], and [iterators]) be placed in? 

Bucket 2: 10

Bucket 3: 1

__Attendance:__ 19

__Author Position:__ Bucket 2

Conclusion: Consensus for Bucket 2.

### Summary

TODO (Nevin)

### Outcome

TODO (Nevin)

**Comment by wg21bot** [comment]
[P1642R5](https://wg21.link/p1642r5) Freestanding Library: Easy [utilities], [ranges], and [iterators] (Ben Craig)

**Comment by brycelelbach** [comment]
# 2021 Winter Library Evolution Polls

[P2333](https://wg21.link/P2333): 2020 Fall Library Evolution Poll Outcomes

__POLL 2: Send P1642R5 (Freestanding [utilities], [ranges], And [iterators]) to LWG, classified as an improvement of an existing feature (P0592R4 bucket 2 item).__

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
|13|13|0|0|0|

__Outcome:__ Unanimous consensus in favor.

### Outcome

Send P1642R5 (Freestanding [utilities], [ranges], And [iterators]) to LWG, classified as an improvement of an existing feature (P0592R4 bucket 2 item).

**Comment by wg21bot** [comment]
[P1642R6](https://wg21.link/p1642r6) Freestanding Library: Easy [utilities], [ranges], and [iterators] (Ben Craig)

**Comment by brycelelbach** [comment]
The 2021 Winter Library Evolution poll did not specify ship vehicle due to a typo, but I believe that C++23 Library Evolution's understanding and desire.

**Comment by wg21bot** [comment]
[P1642R8](https://wg21.link/p1642r8) Freestanding Library: Easy [utilities], [ranges], and [iterators] (Ben Craig)

**Comment by wg21bot** [comment]
[P1642R10](https://wg21.link/p1642r10) Freestanding Library: Easy [utilities], [ranges], and [iterators] (Ben Craig)

**Comment by jensmaurer** [comment]
CWG telecon 2022-07-01: Approved after minor adjustments. LWG to move in plenary.

**Comment by JeffGarland** [comment]
Adding notes from earlier LWG review of R9 on 2022-06-17

https://wiki.edg.com/bin/view/Wg21telecons2022/P1642-20220617

poll: put P1642r10 (to be checked by JG and TK and pending Core approval) into C++23

F | A | N
-- | -- | --
9 | 0 | 0

**Comment by jwakely** [comment]
> poll: put P1642r10 (to be checked by JG and TK and pending Core approval) into C++23

Core requested another change, so the poll will be for P1642R11

**Comment by wg21bot** [comment]
[P1642R11](https://wg21.link/p1642r11) Freestanding Library: Easy [utilities], [ranges], and [iterators] (Ben Craig)

**Comment by frederick-vs-ja** [comment]
I'm sorry for bothering here... Has `<numbers>` ever been discussed?

**Comment by jwakely** [comment]
I think anything involving float-point types was out of scope of this paper and will be addressed later.
